File: app/src/main/java/com/gnest/remember/view/fragments/ListItemFragment.java
package com.gnest.remember.view.fragments;

import android.app.AlarmManager;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.Bundle;
import android.os.Parcelable;
import android.service.notification.StatusBarNotification;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.view.ActionMode;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.GravityCompat;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.target.SimpleTarget;
import com.bumptech.glide.request.transition.Transition;
import com.gnest.remember.App;
import com.gnest.remember.R;

import com.gnest.remember.model.db.data.Memo;
import com.gnest.remember.presenter.IListFragmentPresenter;
import com.gnest.remember.presenter.ListFragmentPresenter;
import com.gnest.remember.services.AlarmReceiver;
import com.gnest.remember.view.IListFragmentView;
import com.gnest.remember.view.activity.MainActivity;
import com.gnest.remember.view.helper.ItemTouchHelperCallback;
import com.gnest.remember.services.AlarmService;
import com.gnest.remember.view.menu.ActionMenu;
import com.gnest.remember.view.layoutmanagers.MyGridLayoutManager;
import com.gnest.remember.view.adapters.MySelectableAdapter;
import com.google.android.material.snackbar.Snackbar;
import com.hannesdorfmann.mosby3.mvp.MvpFragment;
import com.squareup.leakcanary.RefWatcher;

import java.lang.ref.WeakReference;
import java.util.Locale;

import androidx.drawerlayout.widget.DrawerLayout;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindString;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import florent37.github.com.rxlifecycle.RxLifecycle;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.realm.RealmResults;

public class ListItemFragment extends MvpFragment<IListFragmentView, IListFragmentPresenter>
        implements MySelectableAdapter.OnItemActionPerformed, ActionMenu.MenuInteractionHelper, IListFragmentView {

    public static final String BUNDLE_KEY = "Bundle key";
    public static final String EXPANDED_KEY = "Expanded key";
    public static final String ARG_COLUMN_COUNT = "ColumnCount";
    public static final String ARG_MEMO_SIZE = "MemoSize";
    public static final String ARG_MEMO_MARGINS = "MemoMargins";
    private static final String SAVED_LAYOUT_MANAGER = "Saved layout manager";
    private static final String SAVED_STATE_KEY = "Saved state key";

    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    @BindView(R.id.items_fragment)
    LinearLayout layout;
    @BindView(R.id.memo_list)
    RecyclerView recyclerView;
    @BindView(R.id.ItemFragmentToolbar)
    Toolbar toolbar;

    @BindString(R.string.note_archived_message)
    String noteArchivedMessage;
    @BindString(R.string.note_archived_message_1)
    String noteArchivedMessage1;
    @BindString(R.string.note_archived_message_2)
    String noteArchivedMessage2;
    @BindString(R.string.note_removed_message)
    String noteRemovedMessage;
    @BindString(R.string.note_removed_message_1)
    String noteRemovedMessage1;
    @BindString(R.string.note_removed_message_2)
    String noteRemovedMessage2;
    @BindString(R.string.send_memo_intent_title)
    String sendMemoIntentTitle;

    private View mView;
    private Unbinder mUnbinder;
    private int mColumnCount;
    private int mMemoSize;
    private int mMargins;
    private OnListItemFragmentInteractionListener mListener;
    private MySelectableAdapter mAdapter;
    private ItemTouchHelper mItemTouchHelper;
    private ActionMode mActionMode;
    private ActionMenu mActionMenu;
    private MyGridLayoutManager mMyGridLayoutManager;
    private DrawerLayout mDrawerLayout;
    private Bundle mSavedState;


    public static ListItemFragment newInstance(int columnCount, int memoSize, int margins) {
        ListItemFragment fragment = new ListItemFragment();
        Bundle args = new Bundle();
        args.putInt(ARG_COLUMN_COUNT, columnCount);
        args.putInt(ARG_MEMO_SIZE, memoSize);
        args.putInt(ARG_MEMO_MARGINS, margins);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        mActionMenu = new ActionMenu(this);
        RefWatcher refWatcher = App.getRefWatcher();
        refWatcher.watch(this);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        mView = inflater.inflate(R.layout.fragment_item_list, container, false);
        mUnbinder = ButterKnife.bind(this, mView);
        if (getActivity() != null) {
            mDrawerLayout = getActivity().findViewById(R.id.drawer_layout);
        }
        WeakReference<LinearLayout> layoutWeakReference = new WeakReference<>(layout);
        Glide.with(this).load(R.drawable.itemfragment_background_pin_board).into(new SimpleTarget<Drawable>() {
            @Override
            public void onResourceReady(@NonNull Drawable resource, @Nullable Transition<? super Drawable> transition) {
                layoutWeakReference.get().setBackground(resource);
            }
        });
        return mView;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (getArguments() != null) {
            mColumnCount = getArguments().getInt(ARG_COLUMN_COUNT);
            mMemoSize = getArguments().getInt(ARG_MEMO_SIZE);
            mMargins = getArguments().getInt(ARG_MEMO_MARGINS);
        }
        Context context = mView.getContext();
        mMyGridLayoutManager = new MyGridLayoutManager(context, mColumnCount);
        recyclerView.setLayoutManager(mMyGridLayoutManager);
        mAdapter = new MySelectableAdapter(mMemoSize, mMargins);
        mAdapter.setActionListener(this);
        mMyGridLayoutManager.setExpandListener(mAdapter);

        recyclerView.setAdapter(mAdapter);
        ItemTouchHelper.Callback callback = new ItemTouchHelperCallback(mAdapter);
        mItemTouchHelper = new ItemTouchHelper(callback);
        mItemTouchHelper.attachToRecyclerView(recyclerView);

        presenter.loadData();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        AppCompatActivity activity = ((AppCompatActivity) getActivity());
        if (activity != null) {
            activity.setSupportActionBar(toolbar);
            ActionBar actionBar = activity.getSupportActionBar();
            if (actionBar != null) {
                actionBar.setDisplayHomeAsUpEnabled(true);
                actionBar.setHomeButtonEnabled(true);
                mListener.syncDrawerToggleState();
            }
        }
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.ab_itemfragment, menu);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnListItemFragmentInteractionListener) {
            mListener = (OnListItemFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnListItemFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
        if (mAdapter != null) {
            mAdapter.setActionListener(null);
        }
        if (mMyGridLayoutManager != null) {
            mMyGridLayoutManager.setExpandListener(null);
        }
        compositeDisposable.dispose();
        shutDownActionMode();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mUnbinder.unbind();
    }


    @Override
    @NonNull
    public IListFragmentPresenter createPresenter() {
        return new ListFragmentPresenter();
    }


    @Override
    public void setData(RealmResults<Memo> data) {
        mAdapter.setMemos(data);
        mAdapter.notifyDataSetChanged();
        checkStateRestore();
    }

    private void checkStateRestore() {
        Bundle bundle = null;
        //Restoring state after config change
        if (mSavedState != null) {
            bundle = mSavedState;
        }
        //Restoring state after returning from edit mode
        else if (getArguments() != null) {
            bundle = getArguments().getBundle(BUNDLE_KEY);
        }
        if (bundle != null) {
            restoreLayoutManagerState(bundle.getInt(MyGridLayoutManager.LM_SCROLL_ORIENTATION_KEY),
                    bundle.getBoolean(EXPANDED_KEY),
                    bundle.getParcelable(SAVED_LAYOUT_MANAGER),
                    bundle.getInt(MyGridLayoutManager.POSITION_KEY));
        }
    }

    private void restoreLayoutManagerState(int orientation, boolean isItemsExtended, @Nullable Parcelable state, int anchorPosition) {
        if (orientation == LinearLayoutManager.HORIZONTAL) {
            mMyGridLayoutManager.setSpanCount(1);
        }
        mMyGridLayoutManager.setOrientation(orientation);
        mAdapter.setItemsExpanded(isItemsExtended);
        if (state != null) {
            mMyGridLayoutManager.onRestoreInstanceState(state);
        }
        mMyGridLayoutManager.scrollToPosition(anchorPosition);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        Bundle state = saveListState();
        outState.putBundle(SAVED_STATE_KEY, state);
    }

    @NonNull
    public Bundle saveListState() {
        mSavedState = new Bundle();
        mSavedState.putParcelable(SAVED_LAYOUT_MANAGER, mMyGridLayoutManager.onSaveInstanceState());
        mSavedState.putBoolean(EXPANDED_KEY, mAdapter.isItemsExpanded());
        mSavedState.putInt(MyGridLayoutManager.LM_SCROLL_ORIENTATION_KEY, mMyGridLayoutManager.getOrientation());
        mSavedState.putInt(MyGridLayoutManager.POSITION_KEY, mMyGridLayoutManager.getLastPosition());
        return mSavedState;
    }

    @Override
    public void onViewStateRestored(@Nullable Bundle savedInstanceState) {
        super.onViewStateRestored(savedInstanceState);
        if (savedInstanceState != null) {
            mSavedState = savedInstanceState.getParcelable(SAVED_STATE_KEY);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.add:
                mListener.onAddButtonPressed();
                return true;
            case android.R.id.home:
                mDrawerLayout.openDrawer(GravityCompat.START);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onArchiveButtonPressed() {
        presenter.processArchiveActionOnSelected(mAdapter.getSelectedIds());
    }

    @Override
    public void onShareButtonPressed() {
        presenter.processShare(mAdapter.getSelectedIds());
    }

    @Override
    public void onDeleteButtonPressed() {
        presenter.processDeleteSelectedMemos(mAdapter.getSelectedIds());
    }

    @Override
    public void onPerformSwipeDismiss(int memoId, int memoPosition) {
        presenter.processSwipeDismiss(memoId, memoPosition);
    }

    @Override
    public Snackbar getArchiveSnackbar(int numOfNotes) {
        return createAndShowSnackbar(getArchiveActionPluralForm(getPlural(numOfNotes)), numOfNotes);
    }

    @Override
    public Snackbar getDeleteSnackbar(int numOfNotes) {
        return createAndShowSnackbar(getRemoveActionPluralForm(getPlural(numOfNotes)), numOfNotes);
    }

    @NonNull
    private Snackbar createAndShowSnackbar(String pluralForm, int numOfNotes) {
        String text = numOfNotes + " " + pluralForm;
        View view = getView();
        if (view == null) {
            throw new NullPointerException("view == null");
        }
          /*Have to pass a dummy listener, so action button could be visible
        * Undo action is being processed by rxSnackbar in presenter*/
        Snackbar snackbar = Snackbar
                .make(getView(), text, Snackbar.LENGTH_SHORT)
                .setAction(android.R.string.cancel, v -> {
                })
                .setActionTextColor(getResources().getColor(R.color.colorPrimaryDark));
        snackbar.show();
        return snackbar;
    }

    String getArchiveActionPluralForm(int plural) {
        switch (plural) {
            case 2:
                return noteArchivedMessage2;
            case 1:
                return noteArchivedMessage1;
            default:
                return noteArchivedMessage;
        }
    }

    private String getRemoveActionPluralForm(int plural) {
        switch (plural) {
            case 2:
                return noteRemovedMessage2;
            case 1:
                return noteRemovedMessage1;
            default:
                return noteRemovedMessage;
        }
    }

    private int getPlural(int numOfNotes) {
        switch (Locale.getDefault().getLanguage()) {
            case "en":
                return numOfNotes == 1 ? 0 : 1;
            case "ru":
                return (numOfNotes % 10 == 1 && numOfNotes % 100 != 11 ?
                        0 : numOfNotes % 10 >= 2 && numOfNotes % 10 <= 4 &&
                        (numOfNotes % 100 < 10 || numOfNotes % 100 >= 20) ? 1 : 2);
            default:
                return 0;
        }
    }

    @Override
    public void shareMemoText(String memoText) {
        Intent intent = new Intent(Intent.ACTION_SEND);
        intent.setType("text/plain");
        intent.putExtra(Intent.EXTRA_TEXT, memoText);
        Intent chooserIntent = Intent.createChooser(intent, sendMemoIntentTitle);
        mActionMode.finish();
        startActivity(chooserIntent);
    }

    @Override
    public void setAlarm(int memoId, long alarmDate, String notificationText, boolean isAlarmSet, boolean isAlarmMovedToMainScreen) {
        Intent intent = AlarmReceiver.getReceiverIntent(getContext(), memoId, notificationText, alarmDate, isAlarmSet, isAlarmMovedToMainScreen);
        mListener.sendBroadcast(intent);
    }

    @Override
    public void removeAlarm(int memoId) {
        FragmentActivity activity = getActivity();
        if (activity != null) {
            AlarmManager manager = (AlarmManager) activity.getSystemService(Context.ALARM_SERVICE);
            Intent intent = AlarmService.getServiceIntent(activity, null, memoId, true);
            PendingIntent pendingIntent = PendingIntent.getService(activity, memoId, intent, PendingIntent.FLAG_UPDATE_CURRENT);
            if (manager != null) {
                manager.cancel(pendingIntent);
            }
        }
    }

    @Override
    public void removeFromCalendar(int id) {
        mListener.removeFromCalendar(id);
    }

    @Override
    public boolean isNotificationVisible(int id) {
        if (Build.VERSION.SDK_INT >= 23) {
            FragmentActivity activity = getActivity();
            NotificationManager manager = null;
            if (activity != null) {
                manager = (NotificationManager) activity.getSystemService(Context.NOTIFICATION_SERVICE);
            }
            if (manager != null) {
                for (StatusBarNotification notification : manager.getActiveNotifications()) {
                    if (notification.getId() == id) {
                        return true;
                    }
                }
            }
            return false;
        } else {
            Intent notificationIntent = new Intent(getContext(), MainActivity.class);
            PendingIntent pendingIntent = PendingIntent.getActivity(getContext(), id, notificationIntent, PendingIntent.FLAG_NO_CREATE);
            return pendingIntent != null;
        }
    }

    @Override
    public void openFromNotification(int id) {
        Disposable disposable = RxLifecycle.with(getLifecycle())
                .onResume()
                .subscribe(event -> presenter.processOpenFromNotification(id));
        compositeDisposable.add(disposable);
    }

    @Override
    public void closeNotification(int id) {
        NotificationManager notificationManager = (NotificationManager) App.self().getSystemService(Context.NOTIFICATION_SERVICE);
        if (notificationManager != null) {
            notificationManager.cancel(id);
        }
    }

    @Override
    public void swapMemos(int fromId, int fromPosition, int toId, int toPosition) {
        presenter.processMemoSwap(fromId, fromPosition, toId, toPosition);
    }

    @Override
    public void onStartDrag(RecyclerView.ViewHolder viewHolder) {
        mItemTouchHelper.startDrag(viewHolder);
    }

    @Override
    public ActionMode getActionMode() {
        return mActionMode;
    }

    @Override
    public void showActionMode() {
        if (mActionMenu != null && getActivity() != null) {
            mActionMode = ((AppCompatActivity) getActivity()).startSupportActionMode(mActionMenu);
        }
    }

    @Override
    public void shutDownActionMode() {
        if (mActionMode != null) {
            mActionMode.finish();
        }
    }

    @Override
    public void updateContextActionMenu(int numOfSelectedItems) {
        if (mActionMode != null) {
            mActionMode.setTitle(String.valueOf(numOfSelectedItems != 0 ? numOfSelectedItems : ""));
        }
    }

    @Override
    public void switchMultiSelect(boolean switchedOn) {
        mAdapter.switchMultiSelect(switchedOn);
    }

    @Override
    public void clearSelection() {
        mAdapter.clearSelectedList();
    }

    @Override
    public void setShareButtonVisibility(boolean isVisible) {
        if (mActionMenu != null) {
            mActionMenu.setShareButtonVisibility(isVisible);
        }
    }

    @Override
    public void onSingleChoiceMemoClicked(Memo memo) {
        presenter.processSingleChoiceClick(memo, LinearLayoutManager.VERTICAL);
    }

    public void onBackButtonPressed() {
        presenter.processPressBackButton(LinearLayoutManager.VERTICAL, LinearLayoutManager.HORIZONTAL, MainActivity.getColumns());
    }

    @Override
    public MySelectableAdapter getAdapter() {
        return mAdapter;
    }

    @Override
    public MyGridLayoutManager getLayoutManager() {
        return mMyGridLayoutManager;
    }

    @Override
    public OnListItemFragmentInteractionListener getInteractionListener() {
        return mListener;
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p/>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnListItemFragmentInteractionListener {
        void onAddButtonPressed();

        void onBackButtonPressed();

        void onEnterEditMode(int memoId);

        void syncDrawerToggleState();

        void sendBroadcast(Intent intent);

        void removeFromCalendar(int id);
    }
}

File: app/src/main/java/com/gnest/remember/ui/fragments/ListItemFragment.kt
package com.gnest.remember.ui.fragments

import android.os.Bundle
import android.os.Parcelable
import androidx.appcompat.view.ActionMode
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.activity.OnBackPressedCallback
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.navigation.fragment.FragmentNavigatorExtras

import com.gnest.remember.App
import com.gnest.remember.R

import com.gnest.remember.services.AlarmService
import com.gnest.remember.ui.layoutmanagers.MyGridLayoutManager
import com.google.android.material.snackbar.Snackbar

import java.util.Locale

import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.gnest.remember.FAB_TRANSITION
import com.gnest.remember.presentation.ui.EditFragmentDirections
import com.gnest.remember.presentation.ui.memolist.MemoItem
import com.gnest.remember.presentation.ui.memolist.MemoBinder
import com.gnest.remember.presentation.viewmodel.EditMemoListViewModel
import io.reactivex.disposables.CompositeDisposable
import kotlinx.android.synthetic.main.fragment_active_list.*
import mva3.adapter.ListSection
import mva3.adapter.MultiViewAdapter
import org.koin.android.viewmodel.ext.android.sharedViewModel

open class ListItemFragment : Fragment() {

    private val compositeDisposable = CompositeDisposable()
    //    private val actionMenu by lazy { ActionMenu(this) }
//    private val listPresenter: IListFragmentPresenter by inject(named(MAIN))
    private val listViewModel: EditMemoListViewModel by sharedViewModel()


    private val backPressedCallback = object : OnBackPressedCallback(true) {
        override fun handleOnBackPressed() {
            onBackPressed()
        }
    }

    open fun onBackPressed() {
        when (myGridLayoutManager?.orientation) {
            RecyclerView.HORIZONTAL -> {
                myGridLayoutManager?.orientation = RecyclerView.VERTICAL
                myGridLayoutManager?.spanCount = mColumnCount
                myGridLayoutManager?.scrollToPosition(myGridLayoutManager?.lastPosition ?: 0)
//                adapter?.isItemsExpanded = false
            }
        }
    }

    private var mColumnCount: Int = 2
    private var mMemoSize: Int = 0
    private var mMargins: Int = 0

    private val adapter: MultiViewAdapter = MultiViewAdapter()
    private val listSections = ListSection<MemoItem>()

    private var itemTouchHelper: ItemTouchHelper? = null
    private var mActionMode: ActionMode? = null
    private var savedState: Bundle? = null

    var myGridLayoutManager: MyGridLayoutManager? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        val refWatcher = App.getRefWatcher()
        refWatcher?.watch(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_active_list, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
//        add.transitionName = ""
//        add.setOnClickListener {
//            val extras = FragmentNavigatorExtras(add to FAB_TRANSITION)
//            findNavController().navigate(EditFragmentDirections.openEdit(), extras)
//        }

        requireActivity().onBackPressedDispatcher.addCallback(this, backPressedCallback)


//        val layoutWeakReference = WeakReference<LinearLayout>(items_fragment)
//        Glide.with(this).load(R.drawable.itemfragment_background_pin_board).into(object : SimpleTarget<Drawable>() {
//            override fun onResourceReady(resource: Drawable, transition: Transition<in Drawable>?) {
//                layoutWeakReference.get()?.background = resource
//            }
//        })
//        getRenderParams()?.let {
//            mColumnCount = it.columns
//            mMemoSize = it.memoSizePx
//            mMargins = it.marginsPx
//
//        }
//        setSupportActionBar(itemFragmentToolbar)
//        supportActionBar()?.setDisplayHomeAsUpEnabled(true)
//        supportActionBar()?.setHomeButtonEnabled(true)
        myGridLayoutManager = MyGridLayoutManager(requireContext(), mColumnCount)
        val layoutManager = GridLayoutManager(requireContext(), 2)
        adapter.setSpanCount(2)
        layoutManager.spanCount = 2
        layoutManager.spanSizeLookup = adapter.spanSizeLookup
        listSections.setSpanCount(2)
        memoActiveList.layoutManager = layoutManager
//        adapter = MemoListAdapter(
//                onItemClick = { item, view ->
////                    val extras = FragmentNavigatorExtras(view to MEMO_TRANSITION)
//                    val action = EditFragmentDirections.openEdit()
//                    action.memoId = item.id
//                    findNavController().navigate(action)
//                },
//                transitionStartListener = { })
//        adapter?.setActionListener(this)
//        myGridLayoutManager?.setExpandListener(adapter)


        adapter.registerItemBinders(MemoBinder())
        adapter.addSection(listSections)

        memoActiveList.adapter = adapter


//        val callback = ItemTouchHelperCallback(adapter)
//        itemTouchHelper = ItemTouchHelper(callback)
//        itemTouchHelper?.attachToRecyclerView(memoList)


        subscribeToListUpdate()


//        if (!checkNotification()) {
//            presenter.loadData()
//        }
    }

    private fun subscribeToListUpdate() {
        listViewModel.list
                .observe(viewLifecycleOwner, Observer<List<MemoItem>> {
                    listSections.addAll(it)

                    //                    adapter?.memos = it
//                    adapter?.notifyDataSetChanged()
//                    checkStateRestore()
                })
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
//        inflater.inflate(R.menu.ab_itemfragment, menu)
    }

    override fun onDetach() {
        super.onDetach()
//        if (adapter != null) {
//            adapter!!.setActionListener(null)
//        }
        if (myGridLayoutManager != null) {
            myGridLayoutManager!!.setExpandListener(null)
        }
        compositeDisposable.dispose()
//        shutDownActionMode()
    }

    override fun onDestroyView() {
        super.onDestroyView()
    }


//    override fun createPresenter(): IListFragmentPresenter = listPresenter


//    override fun setData(data: RealmResults<Memo>) {
//        adapter?.memos = data
//        adapter?.notifyDataSetChanged()
//        checkStateRestore()
//    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        val state = saveListState()
        outState.putBundle(SAVED_STATE_KEY, state)
    }

    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        super.onViewStateRestored(savedInstanceState)
        if (savedInstanceState != null) {
            savedState = savedInstanceState.getParcelable(SAVED_STATE_KEY)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return true
//        return when (item.itemId) {
//            R.id.add -> {
//                openEdit()
//                true
//            }
//            android.R.id.home -> {
//                activity?.drawerLayout?.openDrawer(GravityCompat.START)
//                true
//            }
//            else -> super.onOptionsItemSelected(item)
//        }
    }

//    override fun onArchiveButtonPressed() {
//        presenter.processArchiveActionOnSelected(adapter!!.selectedIds)
//    }
//
//    override fun onShareButtonPressed() {
//        presenter.processShare(adapter!!.selectedIds)
//    }
//
//    override fun onDeleteButtonPressed() {
//        presenter.processDeleteSelectedMemos(adapter!!.selectedIds)
//    }
//
//    override fun onPerformSwipeDismiss(memoId: Int, memoPosition: Int) {
//        presenter.processSwipeDismiss(memoId, memoPosition)
//    }

//    override fun getArchiveSnackbar(numOfNotes: Int): Snackbar {
//        return createAndShowSnackbar(getArchiveActionPluralForm(getPlural(numOfNotes)), numOfNotes)
//    }
//
//    override fun getDeleteSnackbar(numOfNotes: Int): Snackbar {
//        return createAndShowSnackbar(getRemoveActionPluralForm(getPlural(numOfNotes)), numOfNotes)
//    }
//
//    override fun shareMemoText(memoText: String) {
//        val intent = Intent(Intent.ACTION_SEND)
//        intent.type = "text/plain"
//        intent.putExtra(Intent.EXTRA_TEXT, memoText)
//        val chooserIntent = Intent.createChooser(intent, getString(R.string.send_memo_intent_title))
//        mActionMode!!.finish()
//        startActivity(chooserIntent)
//    }
//
//    override fun setAlarm(memoId: Int, alarmDate: Long, notificationText: String, isAlarmSet: Boolean, isAlarmMovedToMainScreen: Boolean) {
//        val intent = AlarmReceiver.getReceiverIntent(context, memoId, notificationText, alarmDate, isAlarmSet, isAlarmMovedToMainScreen)
//        activity?.sendBroadcast(intent)
//    }
//
//    override fun removeAlarm(memoId: Int) {
//        val activity = activity
//        if (activity != null) {
//            val manager = activity.getSystemService(Context.ALARM_SERVICE) as AlarmManager
//            val intent = AlarmService.getServiceIntent(activity, null, memoId, true)
//            val pendingIntent = PendingIntent.getService(activity, memoId, intent, PendingIntent.FLAG_UPDATE_CURRENT)
//            manager.cancel(pendingIntent)
//        }
//    }
//
//    override fun removeFromCalendar(id: Int) {
//        (requireActivity() as? MainActivity)?.removeFromCalendar(id)
//    }
//
//    override fun isNotificationVisible(id: Int): Boolean {
//        if (Build.VERSION.SDK_INT >= 23) {
//            val activity = activity
//            var manager: NotificationManager? = null
//            if (activity != null) {
//                manager = activity.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
//            }
//            if (manager != null) {
//                for (notification in manager.activeNotifications) {
//                    if (notification.id == id) {
//                        return true
//                    }
//                }
//            }
//            return false
//        } else {
//            val notificationIntent = Intent(context, MainActivity::class.java)
//            val pendingIntent = PendingIntent.getActivity(context, id, notificationIntent, PendingIntent.FLAG_NO_CREATE)
//            return pendingIntent != null
//        }
//    }
//
//    override fun openFromNotification(id: Int) {
//        presenter.processOpenFromNotification(id)
//    }
//
//    override fun closeNotification(id: Int) {
//        val notificationManager = requireContext().getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
//        notificationManager.cancel(id)
//    }
//
//    override fun swapMemos(fromId: Int, fromPosition: Int, toId: Int, toPosition: Int) {
//        presenter.processMemoSwap(fromId, fromPosition, toId, toPosition)
//    }

//    override fun onStartDrag(viewHolder: RecyclerView.ViewHolder) {
//        itemTouchHelper?.startDrag(viewHolder)
//    }

//    override fun getActionMode(): ActionMode? {
//        return mActionMode
//    }

//    override fun showActionMode() {
//        mActionMode = (activity as? AppCompatActivity)?.startSupportActionMode(actionMenu)
//    }

//    override fun shutDownActionMode() {
//        mActionMode?.finish()
//    }

//    override fun updateContextActionMenu(numOfSelectedItems: Int) {
//        mActionMode?.title = when {
//            numOfSelectedItems != 0 -> numOfSelectedItems.toString()
//            else -> ""
//        }
//    }

//    override fun switchMultiSelect(switchedOn: Boolean) {
//        adapter?.switchMultiSelect(switchedOn)
//    }
//
//    override fun clearSelection() {
//        adapter?.clearSelectedList()
//    }

//    override fun setShareButtonVisibility(isVisible: Boolean) {
//        actionMenu.setShareButtonVisibility(isVisible)
//    }

//    override fun onSingleChoiceMemoClicked(memo: Memo) {
//        when {
//            myGridLayoutManager?.orientation == RecyclerView.VERTICAL -> myGridLayoutManager?.openItem(memo.position)
//            else -> {
//                openEdit(memo.id)
//            }
//        }
//    }

//    override fun getAdapter(): MySelectableAdapter? {
//        return adapter
//    }
//
//    override fun getLayoutManager(): MyGridLayoutManager? {
//        return myGridLayoutManager
//    }


    open fun getArchiveActionPluralForm(plural: Int): String {
        return when (plural) {
            2 -> getString(R.string.note_archived_message_2)
            1 -> getString(R.string.note_archived_message_1)
            else -> getString(R.string.note_archived_message)
        }
    }

    private fun checkNotification(): Boolean {
//        requireActivity().intent?.apply {
//            val id = getIntExtra(AlarmService.NOTIFICATION_MEMO_ID, -1)
//            return if (id != -1) {
//                val showOnCurrentScreen = getBooleanExtra(AlarmService.IS_ON_MAIN_SCREEN, true)
//                removeExtra(AlarmService.IS_ON_MAIN_SCREEN)
//                if (showOnCurrentScreen) {
//                    presenter.loadData()
//                } else {
//                    findNavController().navigate(R.id.archive)
//                }
//                removeFromCalendar(id)
//                true
//            } else {
//                false
//            }
//        }
        return false
    }

    private fun checkStateRestore() {
        //Restoring state after config change
        savedState?.let {
            restoreLayoutManagerState(it.getInt(MyGridLayoutManager.LM_SCROLL_ORIENTATION_KEY),
                    it.getBoolean(EXPANDED_KEY),
                    it.getParcelable(SAVED_LAYOUT_MANAGER),
                    it.getInt(MyGridLayoutManager.POSITION_KEY))
            return
        }

        //Restoring state after returning from edit mode
        arguments?.let {
//            ListItemFragmentArgs.fromBundle(it).apply {
//                if (shouldRestore) {
//                    restoreLayoutManagerState(orientation, isExpanded, anchorPosition = position)
//                    return
//                }
//            }
        }

        requireActivity().intent?.apply {
            val id = getIntExtra(AlarmService.NOTIFICATION_MEMO_ID, -1)
            if (id != -1) {
                removeExtra(AlarmService.NOTIFICATION_MEMO_ID)
//                openFromNotification(id)
            }
        }
    }

    private fun restoreLayoutManagerState(orientation: Int, isItemsExtended: Boolean, state: Parcelable? = null, anchorPosition: Int) {
        if (orientation == LinearLayoutManager.HORIZONTAL) {
            myGridLayoutManager?.spanCount = 1
        }
        myGridLayoutManager?.orientation = orientation
//        adapter?.isItemsExpanded = isItemsExtended
        state?.let { myGridLayoutManager?.onRestoreInstanceState(it) }
        myGridLayoutManager?.scrollToPosition(anchorPosition)
    }

    private fun createAndShowSnackbar(pluralForm: String?, numOfNotes: Int): Snackbar {
        val text = "$numOfNotes $pluralForm"
/*Have to pass a dummy listener, so action button could be visible
        * Undo action is being processed by rxSnackbar in presenter*/
        val snackbar = Snackbar
                .make(requireView(), text, Snackbar.LENGTH_SHORT)
                .setAction(android.R.string.cancel) { }
                .setActionTextColor(ContextCompat.getColor(requireContext(), R.color.colorPrimaryDark))
        snackbar.show()
        return snackbar
    }

    private fun getRemoveActionPluralForm(plural: Int): String? {
        return when (plural) {
            2 -> getString(R.string.note_removed_message_2)
            1 -> getString(R.string.note_removed_message_1)
            else -> getString(R.string.note_removed_message)
        }
    }

    private fun getPlural(numOfNotes: Int): Int {
        return when (Locale.getDefault().language) {
            "en" -> if (numOfNotes == 1) 0 else 1
            "ru" -> if (numOfNotes % 10 == 1 && numOfNotes % 100 != 11)
                0
            else if (numOfNotes % 10 in 2..4 &&
                    (numOfNotes % 100 < 10 || numOfNotes % 100 >= 20))
                1
            else
                2
            else -> 0
        }
    }

    private fun saveListState(): Bundle {
        return Bundle().apply {
            putParcelable(SAVED_LAYOUT_MANAGER, myGridLayoutManager?.onSaveInstanceState())
//            putBoolean(EXPANDED_KEY, adapter?.isItemsExpanded ?: false)
            putInt(MyGridLayoutManager.LM_SCROLL_ORIENTATION_KEY, myGridLayoutManager?.orientation
                    ?: RecyclerView.VERTICAL)
            putInt(MyGridLayoutManager.POSITION_KEY, myGridLayoutManager?.lastPosition ?: 0)
            savedState = this
        }
    }


    private fun openEdit(id: Int = -1) {
//        val direction = ListItemFragmentDirections.openEdit().setMemoId(id)
//        findNavController().navigate(direction)
    }

    companion object {

        const val EXPANDED_KEY = "Expanded key"
        private const val SAVED_LAYOUT_MANAGER = "Saved layout manager"
        private const val SAVED_STATE_KEY = "Saved state key"

    }
}